## Chapter 4 – Exercise 2: Sử dụng Mysql Connector để thực hiện các truy vấn dữ liệu đơn giản trên 1 bảng

#### Học viên thực hiện các yêu cầu sau và đối chiếu với kết quả cho trước:

In [1]:
from mysql.connector import connect

In [3]:
# Câu 1: Sử dụng tập tin SALESDB_data.sql để import dữ liệu vào các bảng tương ứng trong database SALESDB
username='root'
pwd=''
server='localhost'
db='salesdb'

connection = connect(host=server, database=db, user=username, password=pwd, port=3306)
curs = connection.cursor()

In [5]:
# Câu 2: Liệt kê danh sách các khách hàng ở Brazil, duyệt và xuất ra từng dòng dữ liệu thỏa điều kiện
curs = connection.cursor()

sql_cmd2 = "select * from customers where Country='Brazil'"

curs.execute(sql_cmd2)

rows2 = curs.fetchall()
for row in rows2:
    print(row)

curs.close()

(15, 'Pedro', 'Afonso', 'Sao Paulo', 'Brazil', '(11) 555-7647')
(21, 'Aria', 'Cruz', 'Sao Paulo', 'Brazil', '(11) 555-9857')
(31, 'André', 'Fonseca', 'Campinas', 'Brazil', '(11) 555-9482')
(34, 'Mario', 'Pontes', 'Rio de Janeiro', 'Brazil', '(21) 555-0091')
(61, 'Bernardo', 'Batista', 'Rio de Janeiro', 'Brazil', '(21) 555-4252')
(62, 'Lúcia', 'Carvalho', 'Sao Paulo', 'Brazil', '(11) 555-1189')
(67, 'Janete', 'Limeira', 'Rio de Janeiro', 'Brazil', '(21) 555-3412')
(81, 'Anabela', 'Domingues', 'Sao Paulo', 'Brazil', '(11) 555-2167')
(88, 'Paula', 'Parente', 'Resende', 'Brazil', '(14) 555-8122')


True

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>(15, 'Pedro', 'Afonso', 'Sao Paulo', 'Brazil', '(11) 555-7647')
(61, 'Bernardo', 'Batista', 'Rio de Janeiro', 'Brazil', '(21) 555-4252')
(62, 'Lúcia', 'Carvalho', 'Sao Paulo', 'Brazil', '(11) 555-1189')
(21, 'Aria', 'Cruz', 'Sao Paulo', 'Brazil', '(11) 555-9857')
(81, 'Anabela', 'Domingues', 'Sao Paulo', 'Brazil', '(11) 555-2167')
(31, 'André', 'Fonseca', 'Campinas', 'Brazil', '(11) 555-9482')
(67, 'Janete', 'Limeira', 'Rio de Janeiro', 'Brazil', '(21) 555-3412')
(88, 'Paula', 'Parente', 'Resende', 'Brazil', '(14) 555-8122')
(34, 'Mario', 'Pontes', 'Rio de Janeiro', 'Brazil', '(21) 555-0091')
</pre>
  
</details>

In [7]:
# Câu 3: Liệt kê danh sách các nhà cung cấp không có số Fax, duyệt và xuất ra từng dòng dữ liệu thỏa điều kiện
curs = connection.cursor()

curs.execute("SELECT * FROM `suppliers` WHERE `Fax` IS NULL;")

row3 = curs.fetchone()
while row3:
    print(row3)
    row3=curs.fetchone()

curs.close()

(1, 'Exotic Liquids', 'Charlotte Cooper', None, 'London', 'UK', '(171) 555-2222', None)
(2, 'New Orleans Ca6/Delights', 'Shelley Burke', None, 'New Orleans', 'USA', '(100) 555-4822', None)
(4, 'Tokyo Traders', 'Yoshi Nagase', None, 'Tokyo', 'Japan', '(03) 3555-5011', None)
(5, "Cooperativa de Quesos 'Las Cabras'", 'Antonio del Valle Saavedra', None, 'Oviedo', 'Spain', '(98) 598 76 54', None)
(6, "Mayumi's", 'Mayumi Ohno', None, 'Osaka', 'Japan', '(06) 431-7877', None)
(8, 'Specialty Biscuits, Ltd.', 'Peter Wilson', None, 'Manchester', 'UK', '(161) 555-4448', None)
(10, 'Refrescos Americanas LTDA', 'Carlos Diaz', None, 'Sao Paulo', 'Brazil', '(11) 555 4640', None)
(11, 'Heli Süßwaren GmbH & Co. KG', 'Petra Winkler', None, 'Berlin', 'Germany', '(010) 9984510', None)
(12, 'Plutzer Lebensmittelgroßmärkte AG', 'Martin Bein', None, 'Frankfurt', 'Germany', '(069) 992755', None)
(15, 'Norske Meierier', 'Beate Vileid', None, 'Sandvika', 'Norway', '(0) 2-953010', None)
(16, 'Bigfoot Breweries', 

True

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>(1, 'Exotic Liquids', 'Charlotte Cooper', None, 'London', 'UK', '(171) 555-2222', None)
(2, 'New Orleans Ca6/Delights', 'Shelley Burke', None, 'New Orleans', 'USA', '(100) 555-4822', None)
(4, 'Tokyo Traders', 'Yoshi Nagase', None, 'Tokyo', 'Japan', '(03) 3555-5011', None)
(5, "Cooperativa de Quesos 'Las Cabras'", 'Antonio del Valle Saavedra', None, 'Oviedo', 'Spain', '(98) 598 76 54', None)
(6, "Mayumi's", 'Mayumi Ohno', None, 'Osaka', 'Japan', '(06) 431-7877', None)
(8, 'Specialty Biscuits, Ltd.', 'Peter Wilson', None, 'Manchester', 'UK', '(161) 555-4448', None)
(10, 'Refrescos Americanas LTDA', 'Carlos Diaz', None, 'Sao Paulo', 'Brazil', '(11) 555 4640', None)
(11, 'Heli Süßwaren GmbH &amp; Co. KG', 'Petra Winkler', None, 'Berlin', 'Germany', '(010) 9984510', None)
(12, 'Plutzer Lebensmittelgroßmärkte AG', 'Martin Bein', None, 'Frankfurt', 'Germany', '(069) 992755', None)
(15, 'Norske Meierier', 'Beate Vileid', None, 'Sandvika', 'Norway', '(0) 2-953010', None)
(16, 'Bigfoot Breweries', 'Cheryl Saylor', None, 'Bend', 'USA', '(503) 555-9931', None)
(17, 'Svensk Sjöföda AB', 'Michael Björn', None, 'Stockholm', 'Sweden', '08-123 45 67', None)
(20, 'Leka Trading', 'Chandra Leka', None, 'Singapore', 'Singapore', '555-8787', None)
(23, 'Karkki Oy', 'Anne Heikkonen', None, 'Lappeenranta', 'Finland', '(953) 10956', None)
(25, 'Ma Maison', 'Jean-Guy Lauzon', None, 'Montréal', 'Canada', '(514) 555-9022', None)
(27, 'Escargots Nouveaux', 'Marie Delamare', None, 'Montceau', 'France', '85.57.00.07', None)
</pre>
  
</details>

In [11]:
# Câu 4: Liệt kê danh sách tất cả các sản phẩm có đơn giá từ 30 đến 60 usd
curs = connection.cursor()

sql_cmd4 = '''
SELECT * FROM `products` 
WHERE `UnitPrice` BETWEEN 30 AND 60
ORDER BY `UnitPrice` ASC;
'''

curs.execute(sql_cmd4)

rows4 = curs.fetchall()
for row in rows4:
    print(row)

curs.close()

(7, "Uncle Bob's Organic Dried Pears", 3, Decimal('30.00'), '12 - 1 lb pkgs.', 0)
(26, 'Gumbär Gummibärchen', 11, Decimal('31.23'), '100 - 250 g bags', 0)
(10, 'Ikura', 4, Decimal('31.50'), '12 - 200 ml jars', 0)
(32, 'Mascarpone Fabioli', 14, Decimal('32.00'), '24 - 200 g pkgs.', 0)
(53, 'Perth Pasties', 24, Decimal('32.80'), '48 pieces', 1)
(64, 'Wimmers gute Semmelknödel', 12, Decimal('33.25'), '20 bags x 4 pieces', 0)
(60, 'Camembert Pierrot', 28, Decimal('34.00'), '15 - 300 g rounds', 0)
(72, 'Mozzarella di Giovanni', 14, Decimal('34.80'), '24 - 200 g pkgs.', 0)
(69, 'Gudbrandsdalsost', 15, Decimal('36.00'), '10 kg pkg.', 0)
(12, 'Queso Manchego La Pastora', 5, Decimal('38.00'), '10 - 500 g pkgs.', 0)
(56, 'Gnocchi di nonna Alice', 26, Decimal('38.00'), '24 - 250 g pkgs.', 0)
(17, 'Alice Mutton', 7, Decimal('39.00'), '20 - 1 kg tins', 1)
(8, 'Northwoods Cranberry Sauce', 3, Decimal('40.00'), '12 - 12 oz jars', 0)
(27, 'Schoggi Schokolade', 11, Decimal('43.90'), '100 - 100 g pieces

True

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>(59, 'Raclette Courdavault', 28, Decimal('55.00'), '5 kg pkg.', 0)
(51, 'Manjimup Dried Apples', 24, Decimal('53.00'), '50 - 300 g pkgs.', 0)
(62, 'Tarte au sucre', 29, Decimal('49.30'), '48 pies', 0)
(43, 'Ipoh Coffee', 20, Decimal('46.00'), '16 - 500 g tins', 0)
(28, 'Rössle Sauerkraut', 12, Decimal('45.60'), '25 - 825 g cans', 1)
(27, 'Schoggi Schokolade', 11, Decimal('43.90'), '100 - 100 g pieces', 0)
(63, 'Vegie-spread', 7, Decimal('43.90'), '15 - 625 g jars', 0)
(8, 'Northwoods Cranberry Sauce', 3, Decimal('40.00'), '12 - 12 oz jars', 0)
(17, 'Alice Mutton', 7, Decimal('39.00'), '20 - 1 kg tins', 1)
(12, 'Queso Manchego La Pastora', 5, Decimal('38.00'), '10 - 500 g pkgs.', 0)
(56, 'Gnocchi di nonna Alice', 26, Decimal('38.00'), '24 - 250 g pkgs.', 0)
(69, 'Gudbrandsdalsost', 15, Decimal('36.00'), '10 kg pkg.', 0)
(72, 'Mozzarella di Giovanni', 14, Decimal('34.80'), '24 - 200 g pkgs.', 0)
(60, 'Camembert Pierrot', 28, Decimal('34.00'), '15 - 300 g rounds', 0)
(64, 'Wimmers gute Semmelknödel', 12, Decimal('33.25'), '20 bags x 4 pieces', 0)
(53, 'Perth Pasties', 24, Decimal('32.80'), '48 pieces', 1)
(32, 'Mascarpone Fabioli', 14, Decimal('32.00'), '24 - 200 g pkgs.', 0)
(26, 'Gumbär Gummibärchen', 11, Decimal('31.23'), '100 - 250 g bags', 0)
(10, 'Ikura', 4, Decimal('31.00'), '12 - 200 ml jars', 0)
(7, "Uncle Bob's Organic Dried Pears", 3, Decimal('30.00'), '12 - 1 lb pkgs.', 0)
</pre>
  
</details>

In [14]:
# Câu 5: Liệt kê danh sách tất cả các sản phẩm có đơn giá lớn hơn 60 usd và đã ngưng bán
curs = connection.cursor()

sql_cmd5 = '''
SELECT * FROM `products` 
WHERE `UnitPrice`>60  AND IsDiscontinued=1;
'''

curs.execute(sql_cmd5)

rows5 = curs.fetchall()
for row in rows5:
    print(row)

curs.close()

(9, 'Mishi Kobe Niku', 4, Decimal('97.00'), '18 - 500 g pkgs.', 1)
(29, 'Thüringer Rostbratwurst', 12, Decimal('123.79'), '50 bags x 30 sausgs.', 1)


True

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>(9, 'Mishi Kobe Niku', 4, Decimal('97.00'), '18 - 500 g pkgs.', 1)
(29, 'Thüringer Rostbratwurst', 12, Decimal('123.79'), '50 bags x 30 sausgs.', 1)
</pre>
  
</details>

In [18]:
# Câu 6: Liệt kê danh sách các khách hàng trong tên có 2 chữ 'o' (vd: Antonio)
curs = connection.cursor()

sql_cmd6 = '''
SELECT * FROM `customers` 
WHERE FirstName LIKE '%o%o%' OR LastName LIKE '%o%o%';
'''

curs.execute(sql_cmd6)

rows6 = curs.fetchall()
for row in rows6:
    print(row)

curs.close()

(3, 'Antonio', 'Moreno', 'México D.F.', 'Mexico', '(5) 555-3932')
(6, 'Hanna', 'Moos', 'Mannheim', 'Germany', '0621-08460')
(15, 'Pedro', 'Afonso', 'Sao Paulo', 'Brazil', '(11) 555-7647')
(27, 'Paolo', 'Accorti', 'Torino', 'Italy', '011-4988260')
(66, 'Maurizio', 'Moroni', 'Reggio Emilia', 'Italy', '0522-556721')
(80, 'Miguel', 'Angel Paolino', 'México D.F.', 'Mexico', '(5) 555-2933')
(87, 'Pirkko', 'Koskitalo', 'Oulu', 'Finland', '981-443655')


True

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>(3, 'Antonio', 'Moreno', 'México D.F.', 'Mexico', '(5) 555-3932')
(6, 'Hanna', 'Moos', 'Mannheim', 'Germany', '0621-08460')
(15, 'Pedro', 'Afonso', 'Sao Paulo', 'Brazil', '(11) 555-7647')
(66, 'Maurizio', 'Moroni', 'Reggio Emilia', 'Italy', '0522-556721')
(80, 'Miguel', 'Angel Paolino', 'México D.F.', 'Mexico', '(5) 555-2933')
(87, 'Pirkko', 'Koskitalo', 'Oulu', 'Finland', '981-443655')
</pre>
  
</details>

In [25]:
# Câu 7: Liệt kê danh sách các khách hàng có tên bắt đầu bằng chữ 'A' và kết thúc bằng chữ 'o' hoặc ở thành phố bắt đầu bằng 'Ma'
curs = connection.cursor()

sql_cmd7 = '''
SELECT * FROM `customers` 
WHERE FirstName LIKE 'A%o' OR LastName LIKE 'A%o' 
OR City LIKE 'Ma%';
'''

curs.execute(sql_cmd7)

rows7 = curs.fetchall()
for row in rows7:
    print(row)

curs.close()

(3, 'Antonio', 'Moreno', 'México D.F.', 'Mexico', '(5) 555-3932')
(6, 'Hanna', 'Moos', 'Mannheim', 'Germany', '0621-08460')
(8, 'Martín', 'Sommer', 'Madrid', 'Spain', '(91) 555 22 82')
(9, 'Laurence', 'Lebihan', 'Marseille', 'France', '91.24.45.40')
(15, 'Pedro', 'Afonso', 'Sao Paulo', 'Brazil', '(11) 555-7647')
(22, 'Diego', 'Roel', 'Madrid', 'Spain', '(91) 555 94 44')
(69, 'Alejandra', 'Camino', 'Madrid', 'Spain', '(91) 745 6200')
(80, 'Miguel', 'Angel Paolino', 'México D.F.', 'Mexico', '(5) 555-2933')


True

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>(6, 'Hanna', 'Moos', 'Mannheim', 'Germany', '0621-08460')
(8, 'Martín', 'Sommer', 'Madrid', 'Spain', '(91) 555 22 82')
(9, 'Laurence', 'Lebihan', 'Marseille', 'France', '91.24.45.40')
(15, 'Pedro', 'Afonso', 'Sao Paulo', 'Brazil', '(11) 555-7647')
(22, 'Diego', 'Roel', 'Madrid', 'Spain', '(91) 555 94 44')
(69, 'Alejandra', 'Camino', 'Madrid', 'Spain', '(91) 745 6200')
(80, 'Miguel', 'Angel Paolino', 'México D.F.', 'Mexico', '(5) 555-2933')
</pre>
  
</details>